Fix _crop and tests

dinkolubina opened the pull request:
Issue being fixed:
DecompressionBombWarning/Error caused by wrong values passed to it in Image._crop

_bomb_decompression_check checks if the size of the image is larger than a certain value.
In Image._crop which is called from Image.crop x1 and y1 are passed to check the size of the crop.
abs(x1 - x0) and abs(y1-y0) should be passed instead, since that represents the size of the crop not just x1 and y1.

Changes proposed in this pull request:
* Use absolute value of (x1-x0) and (y1-y0) instead of the existing check
* Pas those values which represent the width and height instead just the location of x1, y1, which is not correct in the context of _decompression_bomb_check


Pull request is solving this issue: https://github.com/python-pillow/Pillow/issues/3312

hugovk commented on Tests/test_image_crop.py:
```diff
@@ -103,6 +103,30 @@ def test_crop_zero(self):
         self.assertEqual(cropped.size, (10, 10))
         self.assertEqual(cropped.getdata()[2], (0, 0, 0))
 
+    def test_crop_decompression_checks(self):
```
Maybe this test case would be better in `Tests/test_decompression_bomb.py`?

hugovk commented on src/PIL/Image.py:
```diff
@@ -1104,12 +1104,9 @@ def _crop(self, im, box):
-            x1 = x0
-        if y1 < y0:
-            y1 = y0
+        absolute_values = (abs(x1-x0), abs(y1-y0))
```
The outer parentheses can be removed (and spaces added around the minuses).

dinkolubina commented on Tests/test_decompression_bomb.py:
```diff
@@ -61,6 +62,28 @@ def testEnlargeCrop(self):
         self.assert_warning(Image.DecompressionBombWarning,
                             self.src.crop, box)
 
+    def test_crop_decompression_checks(self):
```
Moved the test to test_decompression_bomb.py

dinkolubina commented on src/PIL/Image.py:
```diff
@@ -1104,12 +1104,9 @@ def _crop(self, im, box):
-            x1 = x0
-        if y1 < y0:
-            y1 = y0
+        absolute_values = (abs(x1 - x0), abs(y1 - y0))
```
left the parentheses since _decompression_bomb_check expects a tuple (Image.size), added spaces around minuses

hugovk replied:
It's a tuple with or without parentheses:

```
⌂86% [hugo:~/github/pypistats] master* ± p
Python 3.7.0 (default, Jun 29 2018, 20:13:13)
[Clang 9.1.0 (clang-902.0.39.2)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> absolute_values = (1, 2)
>>> type(absolute_values)
<class 'tuple'>
>>> absolute_values = 1, 2
>>> type(absolute_values)
<class 'tuple'>
>>>
⌂140% [hugo:~/github/pypistats] master* 21s ± p2
Python 2.7.15 (default, Jun 17 2018, 12:46:58)
[GCC 4.2.1 Compatible Apple LLVM 9.1.0 (clang-902.0.39.2)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> absolute_values = (1, 2)
>>> type(absolute_values)
<type 'tuple'>
>>> absolute_values = 1, 2
>>> type(absolute_values)
<type 'tuple'>
>>>
```

dinkolubina replied:
Thanks for the suggestion, didn't know that.

dinkolubina commented on Tests/test_decompression_bomb.py:
```diff
@@ -61,6 +62,28 @@ def testEnlargeCrop(self):
+            self.assertEqual(im.crop(value).size, (9,9))
+
+        for value in warning_values:
+            self.assert_warning(DecompressionBombWarning, im.crop, value)
```
Changed to `assert_warning` instead `assertWarns` as per instructions to stop the test from failing in Python 2.

dinkolubina commented on Tests/test_decompression_bomb.py:
```diff
@@ -61,6 +62,28 @@ def testEnlargeCrop(self):
+            self.assert_warning(DecompressionBombWarning, im.crop, value)
+
+        for value in error_values:
+            with self.assertRaises(DecompressionBombError):
```
Changed per instructions (moved the assertRaises in the for loop)

dinkolubina commented:
I rebased after changing the code. Didn't know that the comments will be deleted, sorry about that. I have addressed the comments individually in the recent commit.
